web/modules/custom/custom_page/src/Plugin/Block/TableOfContentsOutline.php

```diff
@@ -0,0 +1,368 @@
+   *
+   * To get the correct element we will use the flattened list in which holds
+   * the natural prev/next list.
+   *
+   * @return \Drupal\Core\Menu\MenuLinkInterface|null
+   *   The next menu link.
+   */
+  protected function getNextElement(): ?MenuLinkInterface {
+    $flattened_menu = $this->getFlattenedMenu();
+    $active_link_id = $this->getActiveLink()->getPluginId();
+
+    $next = NULL;
+    reset($flattened_menu);
+    do {
+      if (($key = key($flattened_menu)) && $key === $active_link_id) {
+        return next($flattened_menu) ? current($flattened_menu) : NULL;
+      }
+    } while (next($flattened_menu));
+    return NULL;
+  }
+
+  /**
+   * Returns the parent element related to the active link.
+   *
+   * Recursively check the structured menu tree in order to obtain the parent
+   * link.
+   *
+   * @param \Drupal\Core\Menu\MenuLinkTreeElement[] $tree
+   *   The menu tree or subtree to check for the parent.
+   * @param string $active_link_id
+   *   The active link id.
+   * @param \Drupal\Core\Menu\MenuLinkTreeElement $parent
+   *   The current parent found.
+   *
+   * @return \Drupal\Core\Menu\MenuLinkTreeElement|null
+   *   The menu link tree.
+   */
+  protected function getParentElement(array $tree, string $active_link_id, MenuLinkTreeElement $parent = NULL): ?MenuLinkTreeElement {
+    if (empty($tree)) {
+      return NULL;
+    }
+
+    if (isset($tree[$active_link_id])) {
+      return $parent;
+    }
+
+    foreach ($tree as $menu_link_id => $menu_item) {
+      if ($parent = $this->getParentElement($tree[$menu_link_id]->subtree, $active_link_id, $menu_item)) {
+        return $parent;
+      }
+    }
+
+    // If none of the above have found the parent, then the link was not found.
+    return $parent;
+  }
+
+  /**
+   * Loads and returns the og menu instance.
+   *
+   * @return \Drupal\og_menu\OgMenuInstanceInterface
+   *   The og menu instance.
+   */
+  protected function getOgMenuInstance(): OgMenuInstanceInterface {
+    if (empty($this->ogMenuInstance)) {
+      $this->ogMenuInstance = $this->ogMenuManager->getOgMenuInstanceByCustomPage($this->node);
+    }
+
+    return $this->ogMenuInstance;
+  }
+
+  /**
+   * Returns the active link.
+   *
+   * @return \Drupal\Core\Menu\MenuLinkInterface
+   *   The active link.
+   */
+  protected function getActiveLink(): MenuLinkInterface {
+    if (empty($this->activeLink)) {
+      $og_menu_id = $this->getOgMenuName();
+      $this->activeLink = $this->activeTrail->getActiveLink($og_menu_id);
+    }
+
+    return $this->activeLink;
+  }
+
+  /**
+   * Returns the menu tree of the current og menu.
+   *
+   * @return \Drupal\Core\Menu\MenuLinkTreeElement[]
+   *   The og menu tree.
+   */
+  protected function getMenuTree(): array {
+    if (empty($this->menuTree)) {
+      $og_menu_id = $this->getOgMenuName();
+      $this->menuTree = $this->menuLinkTree->load($og_menu_id, new MenuTreeParameters());
```

Comment:
I have added a test that shows that when a menu item links to an unpublished page, it still shows in the menu. We need to run access checks on this.
We also can use the `MenuTreeParameters` to load only the active items. For both you can look at `web/core/modules/system/src/Controller/SystemController.php:106`.
Added a test also for this.